File: ganilib-core/src/main/java/com/gani/lib/ui/view/GSpinner.java
package com.gani.lib.ui.view;

import android.content.Context;
import android.content.res.ColorStateList;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.AppCompatSpinner;
import android.util.AttributeSet;
import android.widget.SpinnerAdapter;

public class GSpinner extends AppCompatSpinner implements IView {
//  private static Spec defaultSpec = new Spec();
//
//  public static void setDefaultSpec(Spec defaultSpec) {
//    GSpinner.defaultSpec = defaultSpec;
//  }

  private ViewHelper helper;

  public GSpinner(Context context) {
    super(context);
    init();
  }

  public GSpinner(Context context, AttributeSet attrs) {
    super(context, attrs);
    init();
  }

  private void init() {
    this.helper = new ViewHelper(this);
//    defaultSpec.init(this);
  }

  public GSpinner adapter(SpinnerAdapter adapter) {
    setAdapter(adapter);
    return this;
  }

  public GSpinner bgResource(int resId) {
    setBackgroundResource(resId);
    return this;
  }

  public GSpinner bgTintList(ColorStateList colorStateList) {
    ViewCompat.setBackgroundTintList(this, colorStateList);
    return this;
  }

  public GSpinner enabled(boolean enabled) {
    setEnabled(enabled);
    return this;
  }

  @Override
  public GSpinner padding(Integer left, Integer top, Integer right, Integer bottom) {
    helper.padding(left, top, right, bottom);
    return this;
  }

  @Override
  public GSpinner margin(Integer left, Integer top, Integer right, Integer bottom) {
    helper.margin(left, top, right, bottom);
    return this;
  }

  public GSpinner onItemSelected(OnItemSelectedListener listener) {
    setOnItemSelectedListener(listener);
    return this;
  }
}

File: ganilib-core/src/main/java/com/glib/core/ui/view/GSpinner.kt
package com.glib.core.ui.view

import android.content.Context
import android.content.res.ColorStateList
import androidx.core.view.ViewCompat
import android.util.AttributeSet
import android.widget.AdapterView
import android.widget.SpinnerAdapter
import androidx.appcompat.widget.AppCompatSpinner

class GSpinner : AppCompatSpinner, IView {
    private val helper = ViewHelper(this)

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    override fun width(width: Int?): GSpinner {
        helper.width(width)
        return this
    }

    override fun height(height: Int?): GSpinner {
        helper.height(height)
        return this
    }

    override fun padding(l: Int?, t: Int?, r: Int?, b: Int?): GSpinner {
        helper.padding(l, t, r, b)
        return this
    }

    override fun margin(l: Int?, t: Int?, r: Int?, b: Int?): GSpinner {
        helper.margin(l, t, r, b)
        return this
    }

    override fun bgColor(color: Int): GSpinner {
        helper.bgColor(color)
        return this
    }

    fun adapter(adapter: SpinnerAdapter): GSpinner {
        setAdapter(adapter)
        return this
    }

    fun bgResource(resId: Int): GSpinner {
        setBackgroundResource(resId)
        return this
    }

    fun bgTintList(colorStateList: ColorStateList): GSpinner {
        ViewCompat.setBackgroundTintList(this, colorStateList)
        return this
    }

    fun enabled(enabled: Boolean): GSpinner {
        isEnabled = enabled
        return this
    }

    fun onItemSelected(listener: AdapterView.OnItemSelectedListener): GSpinner {
        onItemSelectedListener = listener
        return this
    }
}
